modyn/supervisor/internal/triggers/performancetrigger.py

```diff
@@ -0,0 +1,302 @@
+from __future__ import annotations
+
+import logging
+from collections.abc import Generator
+
+from modyn.config.schema.pipeline.trigger.performance.criterion import (
+    DynamicPerformanceThresholdCriterion,
+    StaticNumberAvoidableMisclassificationCriterion,
+    StaticPerformanceThresholdCriterion,
+)
+from modyn.config.schema.pipeline.trigger.performance.performance import (
+    PerformanceTriggerConfig,
+)
+from modyn.evaluator.internal.core_evaluation import perform_evaluation, setup_metrics
+from modyn.evaluator.internal.metrics.accuracy import Accuracy
+from modyn.supervisor.internal.triggers.models import (
+    PerformanceTriggerEvalLog,
+    TriggerPolicyEvaluationLog,
+)
+from modyn.supervisor.internal.triggers.performance.data_density_tracker import (
+    DataDensityTracker,
+)
+from modyn.supervisor.internal.triggers.performance.decision_policy import (
+    DynamicPerformanceThresholdDecisionPolicy,
+    PerformanceDecisionPolicy,
+    StaticNumberAvoidableMisclassificationDecisionPolicy,
+    StaticPerformanceThresholdDecisionPolicy,
+)
+from modyn.supervisor.internal.triggers.performance.performance_tracker import (
+    PerformanceTracker,
+)
+from modyn.supervisor.internal.triggers.trigger import Trigger, TriggerContext
+from modyn.supervisor.internal.triggers.utils.datasets.dataloader_info import (
+    DataLoaderInfo,
+)
+from modyn.supervisor.internal.triggers.utils.datasets.prepare_dataloader import (
+    prepare_trigger_dataloader_fixed_keys,
+)
+from modyn.supervisor.internal.triggers.utils.model.downloader import ModelDownloader
+from modyn.supervisor.internal.triggers.utils.model.manager import ModelManager
+from modyn.utils.utils import LABEL_TRANSFORMER_FUNC_NAME, deserialize_function
+
+logger = logging.getLogger(__name__)
+
+
+class PerformanceTrigger(Trigger):
+    """Trigger based on the performance of the model.
+
+    We support a simple performance drift approach that compares the
+    most recent model performance with an expected performance value
+    that can be static or dynamic through a rolling average.
+
+    Additionally we support a regret based approach where the number of
+    avoidable misclassifications (misclassifications that could have
+    been avoided if we would have triggered) is compared to a threshold.
```

Comment:
I wonder if we should somehow differentiate between these two cases. In the one case, we compare any metric (ROC/AUC, Accuracy, F1...) against a threshold. In the other case, we pick one particular metric (cumulative avoidable misclassifications) and compare that. The only thing that these two have in common is that we compare sth to a threshold. Should we maybe introduce two (sub)classes for these two cases? It could also not make sense, feel free to challenge this (as always), but I feel like we're putting two policies on different levels of abstraction into this class here.

Reply:
I think the current setup is quite suitable, as it's similar to the drift setup where we observe different drift metrics, each having its own `DecisionEngine`. Here, we have different `DecisionEngine`s directly in the `PerformancePolicy`. Of course, we could force the user to split it up into different root-level triggers and use the soon-to-come `EnsembleTriggers` to aggregate them. However, those come with one major problem:

Like embedding generation, evaluations are expensive. They require us to maintain/download models, fetch samples, and run inferences. This is a large chunk of the runtime of these PerformanceTriggers, and it is shared across all `DecisionEngines` and their metrics. Therefore I propose to use those nested `DecisionCriterion` instances (for both avoidable misclassification as well as metric threshold approaches)

The subclass difference between the approaches is already reflected through the hierarchies on `DecisionPolicy` and `DecisionCriterion` (pydantic models) level.